src/nc_ssl.c

```diff
@@ -290,6 +290,50 @@ nc_ssl_read(SSL *ssl, void *buf, size_t num) {
         }
     }
 
+    log_debug(LOG_VERB, "read %d initial bytes", bytes_read);
+
+    // OpenSSL reads data in record size blocks (16kb for SSLv3/TLSv1, see
+    // SSL_read man page for details). Sometimes, SSL_read returns data but
+    // there is still more data available from the next block. Keep reading
+    // until there is actually no more data.
+    //
+    // Don't use SSL_pending since it does not reliably work.
+    for (;;) {
+        void *next_buf = buf + bytes_read;
```

Comment:
Can you explain this a little? I don't understand this syntax of `<array>` + `int`?

Reply:
Good question for two reasons: this is confusing and also it's got a bug!

This is pointer arithmetic. It makes next_buf point to the memory address `bytes_read` elements[1] into the array. 
e.g.
```
char buf[5];
char* next_buf = buf + 2;
```
visually looks like this:
```
memory: [0][0][0][0][0]
         ^     ^
         |     |
         |    next_buf
        buf
```

It turns out this is undefined behavior for void* (since void has no associated size). So really I should cast this before doing the arithmetic.

[1] For characters, which are one byte, `next_buf` would point at the address `buf+1`. But for a 32-bit integer array, `next_buf` would point at the address `buf+4` since a 32-bit integer is 4 bytes.